File: CPP/map/Rank Transform of an Array.cpp
class Solution {
public:
    vector<int> arrayRankTransform(vector<int>& arr) {
        unordered_map<int,int>mp;
        vector<int>ans(arr.size());
        vector<int>arr1(arr);
        sort(arr1.begin(),arr1.end());
        int rank =1;
        for(int i=0;i<arr1.size();i++){
            if(i>0 && arr1[i]>arr1[i-1])rank++;
            mp[arr1[i]]=rank;
        }
        for(int i=0;i<arr.size();i++){
            ans[i]=mp[arr[i]];
        }
        return ans;
    }
};

File: Python/dictionary/Rank Transform of an Array.py
class Solution:
    def arrayRankTransform(self, arr: List[int]) -> List[int]:
        ans = [0] * len(arr)
        arr1 = sorted(arr)  # sorted() returns a new sorted list without modifying arr

        mp= {}  #create a dictionary
        rank = 1
        for i in range(len(arr1)):
            if i>0 and arr1[i]>arr1[i-1]:
                rank+=1
            mp[arr1[i]]=rank
        for i in range(len(arr)):
            ans[i]=mp[arr[i]]
        return ans
